[docs] Adds a doc on LoRA support for diffusers (#2086)
huggingface/diffusers: docs/source/en/training/lora.mdx

an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# LoRA Support in Diffusers 

Diffusers support LoRA for Stable Diffusion for faster fine-tuning allowing greater memory efficiency and easier portability. 

Low-Rank Adaption of Large Language Models was first introduced by Microsoft in
[LoRA: Low-Rank Adaptation of Large Language Models](https://arxiv.org/abs/2106.09685) by *Edward J. Hu, Yelong Shen, Phillip Wallis, Zeyuan Allen-Zhu, Yuanzhi Li, Shean Wang, Lu Wang, Weizhu Chen*.

In a nutshell, LoRA allows adapting pretrained models by adding pairs of rank-decomposition weight matrices (called **update marrices**) 
to existing weights and **only** training those newly added weights. This has a couple of advantages:

- Previous pretrained weights are kept frozen so that model is not prone to [catastrophic forgetting](https://www.pnas.org/doi/10.1073/pnas.1611835114). 
- Rank-decomposition matrices have significantly fewer parameters than original model, which means that trained LoRA weights are easily portable.

> - LoRA attention layers allow to control to which extent the model is adapted toward new training images via a `scale` parameter.

Author: See if the current edits make sense.

[cloneofsimo](https://github.com/cloneofsimo) was the first to try out LoRA training for Stable Diffusion in the popular [lora](https://github.com/cloneofsimo/lora) GitHub repository.

<Tip>

LoRA also allows us to achieve greater memory efficiency since the pretrained weights are kept frozen, only the LoRA weights are trained, thereby

> allowing us to run fine-tuning on consumer GPUs like Tesla T4.

Reviewer: Update: text_to_image finetuning works on a 2080 Ti!

```suggestion
allowing us to run fine-tuning on consumer GPUs like Tesla T4, RTX 3080 or even RTX 2080 Ti!
```

> allowing us to run fine-tuning on consumer GPUs like Tesla T4.

Author: Goodness!

  --resolution=512 \
  --train_batch_size=1 \
  --gradient_accumulation_steps=1 \
  --checkpointing_steps=100 \
  --learning_rate=1e-4 \
  --report_to="wandb" \
  --lr_scheduler="constant" \
  --lr_warmup_steps=0 \
  --max_train_steps=500 \
  --validation_prompt="A photo of sks dog in a bucket" \
  --validation_epochs=50 \
  --seed="0" \
  --push_to_hub
```

Refer to the respective examples linked above to learn more. 

<Tip>

When using LoRA we can use a much higher learning rate (typically 1e-4 as opposed to 1e-5) compared to non-LoRA fine-tuning.

</Tip>

But there is no free lunch. For the given dataset and expected generation quality, you'd still need to experiment with
different hyperparameters. Here are some important ones:

* Training time
    * Learning rate 
    * Number of training steps
* Inference time 
    * Number of steps 
    * Scheduler type

Additionally, you can follow [this blog](https://huggingface.co/blog/dreambooth) that documents some of our experimental
findings for performing DreamBooth training Stable Diffusion.

When fine-tuning, the LoRA update matrices are only added to the attention layers. To enable this, we added new weight

> loading functionalities. Their details are available [here](https://huggingface.co/docs/diffusers/main/en/api/loaders).

Reviewer: Is this because of this particular implementation or something general in LoRA? (Just curious, no need to change anything here).

Author: I think the bit on only considering attention layers for adding LoRA weights is coming from the original LoRA paper. There, the authors only experimented with the attention layers.

## Inference 

> Assuming, you used the `examples/text_to_image/train_text_to_image_lora.py` to fine-tune Stable Diffusion on the [Pokemons

Reviewer: ```suggestion
Assuming you used the `examples/text_to_image/train_text_to_image_lora.py` to fine-tune Stable Diffusion on the [Pokemon
```

Reviewer: Also, the previous example was about dreambooth, not fine-tuning. Maybe include a brief mention before this section? Something like `A similar process can be followed to fully fine-tune Stable Diffusion on a custom dataset using the train_text_to_image_lora script`

Reviewer 2: +1

> dataset](https://huggingface.co/lambdalabs/pokemon-blip-captions), you can perform inference like so: 

Reviewer: ```suggestion
dataset](https://huggingface.co/datasets/lambdalabs/pokemon-blip-captions), you can perform inference like so: 
```

```py 
from diffusers import StableDiffusionPipeline
import torch

model_path = "sayakpaul/sd-model-finetuned-lora-t4"

> pipe = StableDiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4", torch_dtype=torch.float16)

Reviewer 3: (nit) maybe we can show how to retrieve the base_model from the model card by loading the yaml code via `huggingface_hub`

Reviewer 3: https://github.com/huggingface/huggingface_hub/blob/5a12851f54bf614be39614034ed3a9031922d297/src/huggingface_hub/repocard.py#L136

Reviewer: Nice! Maybe include it as a tip below the current snippet?

Reviewer 3: For me it's fine in the same code snippet

Author: See if the current changes make sense.

pipe.unet.load_attn_procs(model_path)
pipe.to("cuda")

prompt = "A pokemon with green eyes and red legs."
image = pipe(prompt, num_inference_steps=30, guidance_scale=7.5).images[0]

> image.save("pokemon.png")

Reviewer: Just wondering, maybe display the image here? We never do it in the docs, what's your opinion about starting doing it to make things more visual?

Author: Diffusion for computer vision is definitely about visuals. I like the idea and I think we should definitely add it :)

Author: Added an image.

```

[`sayakpaul/sd-model-finetuned-lora-t4`](https://huggingface.co/sayakpaul/sd-model-finetuned-lora-t4) contains [LoRA fine-tuned update matrices](https://huggingface.co/sayakpaul/sd-model-finetuned-lora-t4/blob/main/pytorch_lora_weights.bin)

> which is only 3 MBs in size. During inference, the pre-trained Stable Diffusion checkpoints loaded alongside these update

Reviewer: ```suggestion
which is only 3 MBs in size. During inference, the pre-trained Stable Diffusion checkpoints are loaded alongside these update
```

an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# LoRA Support in Diffusers 

Diffusers supports LoRA for faster fine-tuning of Stable Diffusion, allowing greater memory efficiency and easier portability. 

Low-Rank Adaption of Large Language Models was first introduced by Microsoft in
[LoRA: Low-Rank Adaptation of Large Language Models](https://arxiv.org/abs/2106.09685) by *Edward J. Hu, Yelong Shen, Phillip Wallis, Zeyuan Allen-Zhu, Yuanzhi Li, Shean Wang, Lu Wang, Weizhu Chen*.

In a nutshell, LoRA allows adapting pretrained models by adding pairs of rank-decomposition weight matrices (called **update matrices**) 
to existing weights and **only** training those newly added weights. This has a couple of advantages:

- Previous pretrained weights are kept frozen so that the model is not so prone to [catastrophic forgetting](https://www.pnas.org/doi/10.1073/pnas.1611835114). 
- Rank-decomposition matrices have significantly fewer parameters than the original model, which means that trained LoRA weights are easily portable.

> - LoRA matrices are generally added to the attention layers of the original model and they control to control to which extent the model is adapted toward new training images via a `scale` parameter.

Reviewer: ```suggestion
- LoRA matrices are generally added to the attention layers of the original model and they control to which extent the model is adapted toward new training images via a `scale` parameter.
```

> **__Note that the usage of LoRA is not limited to only attention layers. In the original LoRA work, the authors found out that just ammending

Reviewer: ```suggestion
**__Note that the usage of LoRA is not just limited to attention layers. In the original LoRA work, the authors found out that just amending
```
